os/drivers/otp/otp.c

```diff
@@ -0,0 +1,240 @@
+ * Name: otp_ioctl
+ *
+ * Description:
+ *    This function is allow you to control the event.
+ *
+ ****************************************************************************/
+static int otp_ioctl(struct file *filep, int cmd, unsigned long arg)
+{
+	return 0;
+}
+
+
+/****************************************************************************
+ * Name: otp_close
+ *
+ * Description:
+ *   This routine is called when the otp gets closed.
+ *
+ ****************************************************************************/
+static int otp_close(struct file *filep)
+{
+	return 0;
+}
+
+/*****************************************************************************
+ * Name: otp_open
+ *
+ * Description:
+ *   This routine is called whenever otp is opened.
+ *
+ ****************************************************************************/
+static int otp_open(struct file *filep)
+{
+	return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: otp_register
+ *
+ * Description:
+ *   Register OTP device.
+ *
+ ****************************************************************************/
+int otp_register(struct otp_lowerhalf_s *lower)
+{
+	int ret;
+	FAR struct otp_upperhalf_s *priv;
+
+	DEBUGASSERT(lower);
+
+	if (g_otp_dev) {
+		return -1;
+	}
+
+	/* Allocate a new GPIO driver instance */
+	priv = (FAR struct otp_upperhalf_s *)
+			kmm_zalloc(sizeof(struct otp_upperhalf_s));
+	if (!priv) {
+		lldbg("ERROR: Failed to allocate device structure\n");
+		return -ENOMEM;
+	}
+
+	/* Initiailize the new GPIO driver instance */
+	priv->ou_lower = lower;
+	lower->parent = priv;
+
+	/* And register the GPIO driver */
+	ret = register_driver(OTP_DEV_PATH, &g_otp_ops, 0666, priv);
+	if (ret < 0) {
+		lldbg("ERROR: register driver failed: %d\n", ret);
+		goto errout_with_priv;
+	}
+	g_otp_dev = priv;
+
+	return OK;
+
+errout_with_priv:
+	sem_destroy(&priv->ou_exclsem);
+	kmm_free(priv);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: otp_unregister
+ *
+ * Description:
+ *   unregister OTP device.
+ *
+ ****************************************************************************/
+int otp_unregister(void)
+{
+	if (!g_otp_dev) {
+		lldbg("ERROR: no registered device\n");
+		return -1;
+	}
+
+	unregister_driver(OTP_DEV_PATH);
+
+	sem_destroy(&g_otp_dev->ou_exclsem);
+	kmm_free(g_otp_dev);
+	g_otp_dev = NULL;
+
+	return 0;
+}
+
+/****************************************************************************
+ * Name: get_otp_dev
+ *
+ * Description:
+ *   Get otp lower device.
+ *   Programs(SE) in a kernel space can access otp by this function
+ *
+ ****************************************************************************/
+otp_dev_s *get_otp_dev(void)
```

Comment:
It is very dangerous approach, I explained to your group many times.
1. Upper never knows which lower is attached.
2. Lower should be controlled strictly, because only upper is registered in iNode structure because lower is private area of upper..
    It means that upper should maintain state of lower too.
    Please assume that there are 2 thread which use this function and read & write simultaneously.
    Each of data will be merged if upper do not lock it.